app/octopus_graphql_api_client.py

```diff
@@ -0,0 +1,173 @@
+import json
+import aiohttp
+from datetime import (datetime, timedelta, time)
+from intelligent_dispatches import IntelligentDispatchItem, IntelligentDispatches
+from contextlib import suppress
+import ciso8601
+
+intelligent_dispatches_query = '''query {{
+  plannedDispatches(accountNumber: "{account_id}") {{
+    startDt
+    endDt
+    delta
+    meta {{
+      source
+      location
+    }}
+  }}
+  completedDispatches(accountNumber: "{account_id}") {{
+    startDt
+    endDt
+    delta
+    meta {{
+      source
+      location
+    }}
+  }}
+}}'''
+
+def parse_datetime(dt_str: str) -> dt.datetime | None:
+  """Parse a string and return a datetime.datetime.
+
+  This function supports time zone offsets. When the input contains one,
+  the output uses a timezone with a fixed offset from UTC.
+  Raises ValueError if the input is well formatted but not a valid datetime.
+  Returns None if the input isn't well formatted.
+  """
+  with suppress(ValueError, IndexError):
+      return ciso8601.parse_datetime(dt_str)
+
+  if not (match := DATETIME_RE.match(dt_str)):
+    return None
+  kws: dict[str, Any] = match.groupdict()
+  if kws["microsecond"]:
+    kws["microsecond"] = kws["microsecond"].ljust(6, "0")
+  tzinfo_str = kws.pop("tzinfo")
+
+  tzinfo: dt.tzinfo | None = None
+  if tzinfo_str == "Z":
+    tzinfo = UTC
+  elif tzinfo_str is not None:
+    offset_mins = int(tzinfo_str[-2:]) if len(tzinfo_str) > 3 else 0
+    offset_hours = int(tzinfo_str[1:3])
+    offset = dt.timedelta(hours=offset_hours, minutes=offset_mins)
+    if tzinfo_str[0] == "-":
+      offset = -offset
+    tzinfo = dt.timezone(offset)
+  kws = {k: int(v) for k, v in kws.items() if v is not None}
+  kws["tzinfo"] = tzinfo
+  return dt.datetime(**kws)
+
+def as_utc(dattim: dt.datetime) -> dt.datetime:
+  """Return a datetime as UTC time.
+
+  Assumes datetime without tzinfo to be in the DEFAULT_TIME_ZONE.
+  """
+  if dattim.tzinfo == UTC:
+    return dattim
+  if dattim.tzinfo is None:
+    dattim = dattim.replace(tzinfo=DEFAULT_TIME_ZONE)
+
+  return dattim.astimezone(UTC)
+
+class OctopusEnergyApiClient:
```

Comment:
I think this class is better named `OctopusGraphQlClient` as there is already `OctopusApiClient` and adding `Energy` to the name doesn't describe that it's the GraphQL client.

Reply:
Again, this isn't my code and I wanted to minimise changes, but I suppose this rename makes sense.